src/jrd/validation.cpp

```diff
@@ -1790,17 +1790,22 @@ Validation::RTN Validation::walk_data_page(jrd_rel* relation, ULONG page_number,
 				// is a backversion then at least one of the record versions is
 				// committed. If there's no backversion then check transaction
 				// state of the lone primary record version. Unless it is already deleted.
+				// Note, if the primary record version is deleted and committed, the
+				// existence of an index entry is not required for such version chain
+				// because it is all garbage.
+				const bool deleted_flag = header->rhd_flags & rhd_deleted;
 
-				if (header->rhd_b_page)
+				if (header->rhd_b_page && !deleted_flag)
 					RBM_SET(pool, &vdr_rel_records, number.getValue());
-				else if ((header->rhd_flags & rhd_deleted) == 0)
+				else if (header->rhd_b_page || !deleted_flag)
```

Comment:
Both ifs contain condition `header->rhd_b_page`. Cannot they be replaced with single `if`?

Reply:
I think I tried it before and it led to unnecessary code duplication.